mantidproject/mantid
# Fix error with GroupingPattern in GroupDetectors

This PR fixes a bug where the GroupingPattern property in the GroupDetectors algorithm does not give the correct results for workspaces where the spectrum number is not 1-based.

**To test:**
Run the following script. Check the output - the last two calls to GroupDetectors should both give the same result (a sum of 19).
```
def printVal(ws, i):
    print "Index=" + str(i) + ", Spectrum=" + str(ws.getSpectrum(i).getSpectrumNo()) + ", Value = " + str(ws.readY(i))

sample = CreateWorkspace(DataX=[1], DataY=[1,2,3,4,5,6,7,8,9,10], NSpec=10)
print "Original workspace:"
for i in range(0,10):
    printVal(sample,i)
print

group1 = GroupDetectors(sample, GroupingPattern="4-5")
print "Grouped original workspace with GroupingPattern='4-5'. This sums rows 4-5 => spectra 5-6 giving a result of 11:"
printVal(group1,0)
print

cropped = CropWorkspace(sample, StartWorkspaceIndex=4)
print "Cropped workspace:"
for i in range(0,6):
    printVal(cropped,i)
print

group2 = GroupDetectors(cropped, GroupingPattern="4-5")
print "Grouped cropped workspace with GroupingPattern='4-5'. This should sum rows 4-5 => spectra 9-10 giving a result of 19 (but in the bug it sums rows 0-1 => spectra 5-6 giving a result of 11):"
printVal(group2,0)
print

group3 = GroupDetectors(cropped, WorkspaceIndexList="4-5")
print "Grouped cropped workspace with WorkspaceIndexList='4-5'. This sums rows 4-5 => spectra 9-10 giving a result of 19:"
printVal(group3,0)
print
```

Fixes #15749 .



---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
